### Step 1: Setup and Data Preparation
1. **Install Required Packages**: Make sure all necessary packages (like `transformers`, `datasets`, etc.) are installed.
2. **Data Acquisition**: Download the dataset mentioned in the README.


In [ ]:
!pip install transformers datasets

Download the dataset as mentioned in the README.

In [4]:
! gdown 1XL3Ob__YnlDaAUWAmsjFeF2zGWn0OdDG

Traceback (most recent call last):
  File "/usr/local/bin/gdown", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/gdown/cli.py", line 151, in main
    filename = download(
  File "/usr/local/lib/python3.10/dist-packages/gdown/download.py", line 203, in download
    filename_from_url = m.groups()[0]
AttributeError: 'NoneType' object has no attribute 'groups'


In [2]:
# import requests
# from zipfile import ZipFile
# from io import BytesIO

# # Replace with the actual URL of the dataset
# url = 'https://drive.google.com/drive/folders/1Euz7DAbiWKAwiTErLBnebbUNkCY0Vo4C'

# response = requests.get(url)
# zip_file = ZipFile(BytesIO(response.content))
# zip_file.extractall('path/to/extract')

<Response [200]>

### Step 2: Create Functions for ChatGPT (GPT-3.5) Inference
1. **Load ChatGPT Model**: Create a function to load the GPT-3.5 model.
2. **Generate Prompts**: Since ChatGPT can't be fine-tuned, we'll rely on prompt engineering. Create a function to format inputs according to the specific recommendation tasks.
3. **Inference Function**: Develop a function to run inference using ChatGPT with the generated prompts.

### Step 3: Adapt Code for LLAMA (Preparing for Future Fine-tuning)
1. **Load LLAMA Model**: Modify the script to load the LLAMA model.
2. **Fine-tuning Setup (Placeholder)**: Prepare the structure for future fine-tuning. This will be a placeholder for now, as actual fine-tuning won't be done in this step.
3. **Inference with LLAMA**: Adapt the inference function for use with the LLAMA model.

### Step 4: Code Cleanup and Documentation
1. **Remove Unnecessary Code**: Identify and remove parts of the code that are not needed (like web interface components).
2. **Add Comments and Documentation**: Ensure each part of the code is well-documented for clarity.

### Step 5: Testing and Final Adjustments (To Be Done by You)
1. **Run the Notebook**: Test each part of the notebook in your environment.
2. **Debug and Tweak**: Make any necessary adjustments based on the test results.





---

Let's start with **Step 1: Setup and Data Preparation**. Here's a basic outline of the code snippets you'll need:

#### Install Required Packages
You can install the necessary packages using pip (if not already installed):
```python
!pip install transformers datasets
```

#### Data Acquisition
Download the dataset as mentioned in the README. The dataset links are provided there. You might need to use the following code to download and extract the data:
```python
import requests
from zipfile import ZipFile
from io import BytesIO

# Replace with the actual URL of the dataset
url = 'DATASET_URL'

response = requests.get(url)
zip_file = ZipFile(BytesIO(response.content))
zip_file.extractall('path/to/extract')
```

Would you like to proceed with creating the functions for ChatGPT inference, or is there any specific aspect you would like to focus on first?

### GPU Setup


for use local gpu run this comand in anaconda prompet:

      1) pip install jupyter_http_over_ws
      2) jupyter serverextension enable --py jupyter_http_over_ws
      3) jupyter notebook --NotebookApp.allow_origin='https://colab.research.google.com' --port=8888 --NotebookApp.port_retries=0
      
copy url and paste on colab

In [ ]:
# import torch
#Check GPU
# torch.cuda.get_device_name(0)

# article

In [ ]:

# from PyPDF2 import PdfReader

# pdf_reader = PdfReader("Is ChatGPT a Good Recommender A Preliminary Study.pdf")
# pdf_text = ''
# for page_num in range(len(pdf_reader.pages)):
#     page = pdf_reader.pages[page_num]
#     pdf_text += page.extract_text()

# pdf_text[401:-15625]

# file_path = 'example.txt'
# with open(file_path, 'w') as file:
#     # 2. Write the string to the file
#     string_to_save = pdf_text[401:-15625]
#     file.write(string_to_save)

This structure indicates that the LLMRec system can handle scenarios where either some input data is provided or where only an instruction is needed. The system then uses these templates to format the prompts sent to the language model, and the responses are extracted based on the specified split marker.

# utils

## Templates `JSON`

In [ ]:
Template = {
    "description": "Template for LLMRec",
    "prompt_input": "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n",
    "prompt_no_input": "### Instruction:\n{instruction}\n\n### Response:\n",
    "response_split": "### Response:"
}


The `rec_template.json` file contains the structure for the prompt templates used in the LLMRec system. Here is a breakdown of its contents:

1. **Description**: It is labeled as 'Template for LLMRec', indicating its purpose in the system.

2. **Prompt with Input (`prompt_input`)**:
   - This template is used when there is an input provided for the prompt.
   - The format includes an 'Instruction' section, an 'Input' section, and a placeholder for 'Response'.
   - The placeholders `{instruction}` and `{input}` will be replaced with actual instruction and input data when generating a prompt.

3. **Prompt without Input (`prompt_no_input`)**:
   - This template is used when there is no input data to include in the prompt.
   - It only includes an 'Instruction' section followed by a placeholder for 'Response'.

4. **Response Split (`response_split`)**:
   - This is a marker used to split the generated text to extract the 'Response' part.
   - It is set to '### Response:', which means this marker is used to identify where the response begins in the model's output.


## Prompter `class`

The design of the `Prompter` class indicates a flexible approach to generating prompts for recommendation tasks, where the specifics of each prompt are determined by a `JSON` template. This aligns with the concept of transforming recommendation tasks into natural language tasks, as described in the article.

In [ ]:
import json
import os.path as osp
from typing import Union

class Prompter(object):

    # Slots for memory efficiency (instances of this class can only have these two attributes, and no other attributes can be added dynamically.)
    __slots__ = ("template", "_verbose")

    def __init__(self, template_name: str = "", verbose: bool = False):
        """
        Initialize the Prompter with a specific template.

        Parameters:
        - template_name (str): The name of the template file to use for generating prompts.
        - verbose (bool): If True, the prompter will print additional information.

        Returns:
        None
        """

        self._verbose = verbose

        # Set a default template if not provided
        if not template_name:
            # Enforce the default here, so the constructor can be called with '' and will not break.
            template_name = "alpaca"
        # Load the template file
        file_name = osp.join("./templates", f"{template_name}.json")
        if not osp.exists(file_name):
            raise ValueError(f"Can't read {file_name}")

        with open(file_name) as fp:
            self.template = json.load(fp)

        if self._verbose:
            print(
                f"Using prompt template {template_name}: {self.template['description']}"
            )

    def generate_prompt(self, instruction: str, input: Union[None, str] = None, label: Union[None, str] = None,) -> str:
        ''''
        Generate a prompt based on the provided instruction and optional input/label

        Parameters:
        - instruction (str): Instruction or context for the prompt.
        - input (Union[None, str]): Optional input data to include in the prompt.
        - label (Union[None, str]): Optional label or expected output to append.

        Returns:
        str: The generated prompt.
        '''

        # if a label (=response, =output) is provided, it's also appended.
        if input:
            res = self.template["prompt_input"].format(
                instruction=instruction, input=input
            )
        else:
            res = self.template["prompt_no_input"].format(
                instruction=instruction
            )

        # Append label (response/output) if provided
        if label:
            res = f"{res}{label}"

        # Print the prompt if verbose mode is on
        if self._verbose:
            print(res)
        return res

    def get_response(self, output: str) -> str:
        """
        Extract the response part from the model's output.

        Parameters:
        - output (str): The complete output text from the model.

        Returns:
        str: The extracted response portion of the output.
        """
        return output.split(self.template["response_split"])[1].strip()


## generation

This will involve extracting the core functionality needed for generating prompts and obtaining model outputs, and then restructuring it for interactive use in a Jupyter notebook.

In [ ]:
! pip install peft

In [ ]:
import sys
import torch
import transformers
from peft import PeftModel # A model wrapper for performance improvements
from transformers import GenerationConfig, LlamaForCausalLM, LlamaTokenizer

# from utils.prompter import Promper # Custom class for managing prompt templates

### load model
Matches and sets up the device for the model, and loads the model accordingly.

In [ ]:
def load_model(base_model: str = "", load_8bit: bool = False, lora_weights: str = "tloen/alpaca-lora-7b"):
  """
  Matches and sets up the device for the model, and loads the model accordingly and config it.

  Parameters:
  - base_model (str): The base model identifier.
  - load_8bit (bool): Whether to load the model in 8-bit precision for better performance.
  - lora_weights (str): Path or identifier for LoRA weights for the model.

  Returns:
  The loaded model, set up for the appropriate device.
  """
  # Ensure that a base model is specified
  assert (
      base_model
  ), "Please specify a --base_model, e.g. --base_model='decapoda-research/llama-7b-hf'"

  # Determine the device: CUDA, MPS, or CPU
  device = "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu"

  # Common parameters for model loading
  common_params = {
    "base_model": base_model,
    "torch_dtype": torch.float16 if device != "cpu" else None,
    "device_map": {"": device}
  }

  # Load the base model with specific settings based on the device
  match device:
    case "cuda":
      common_params.update({"load_in_8bit": load_8bit})
    case "mps":
      pass  # No additional parameters for MPS
    case _:
      common_params.update({"low_cpu_mem_usage": True})

  model = LlamaForCausalLM.from_pretrained(**common_params)
  # Load PeftModel with LoRA weights
  model = PeftModel.from_pretrained(model, lora_weights, **common_params)

  # Unwind broken configurations from a specific source, possibly due to model incompatibilities or custom requirements
  model.config.pad_token_id = tokenizer.pad_token_id = 0  # Set padding token ID to 0 (typically unknown token)
  model.config.bos_token_id = 1  # Set beginning of sentence token ID
  model.config.eos_token_id = 2  # Set end of sentence token ID

  # Convert the model to half precision for improved performance, unless 8-bit loading is specified
  if not load_8bit:
    model.half()  # seems to fix bugs for some users.

  # Set the model to evaluation mode
  model.eval()

  # Compile the model with TorchScript if using PyTorch version 2 or above and not on Windows platform
  if torch.__version__ >= "2" and sys.platform != "win32":
    model = torch.compile(model)
  return model


### generate
Main function to set up and run the language model for generating outputs based on prompts

In [ ]:
# Define the evaluate function to generate responses based on instructions and input
def generate(base_model, instruction, input=None, prompt_template: str = "med_template",
             temperature=0.1, top_p=0.75, top_k=40, num_beams=4, max_new_tokens=128, **kwargs,):
  """
  Generate a response from the model based on the given instruction and optional input.

  This function takes an instruction (and optionally additional input), generates a prompt,
  and then uses the model to generate a response based on this prompt. The function allows
  configuration of various generation parameters like temperature, top_p, top_k, and num_beams.

  Parameters:
  - base_model (str): The base model for language generation to be used.
  - instruction (str): The instruction or query to be processed by the model.
  - input (str, optional): Additional input data to be included in the prompt.
  - temperature (float): The temperature for controlling randomness in response generation.
  - top_p (float): Nucleus sampling parameter controlling the size of the probability mass considered for sampling.
  - top_k (int): Top-k sampling parameter controlling the number of highest probability vocabulary tokens to keep for sampling.
  - num_beams (int): Number of beams for beam search. More beams increase the chances of finding more diverse solutions but are slower.
  - max_new_tokens (int): The maximum number of new tokens to generate in the response.
  - prompt_template (str): Name of the prompt template to use.
  - **kwargs: Additional keyword arguments for generation configuration.

  Returns:
  str: The generated response from the model.
  """
  # Initialize the prompter with the specified template
  prompter = Prompter(prompt_template)

  # Load the tokenizer for the specified base model
  tokenizer = LlamaTokenizer.from_pretrained(base_model)

  # Generate the prompt using the provided instruction and input
  prompt = prompter.generate_prompt(instruction, input)

  # Tokenize the prompt for model input
  inputs = tokenizer(prompt, return_tensors="pt")

  input_ids = inputs["input_ids"].to(device)

  # Configure generation parameters (temperature, top_p, etc.)
  generation_config = GenerationConfig(
    temperature=temperature,
    top_p=top_p,
    top_k=top_k,
    num_beams=num_beams,
    **kwargs,
  )

  # Load the model based on the device and settings
  model = load_model(base_model=base_model)

  # Generate the output without computing gradients (for efficiency)
  with torch.no_grad():
    generation_output = model.generate(
      input_ids=input_ids,
      generation_config=generation_config,
      return_dict_in_generate=True,
      output_scores=True,
      max_new_tokens=max_new_tokens,
    )

  # Decode the generated sequence to human-readable text
  s = generation_output.sequences[0]
  output = tokenizer.decode(s)
  # Extract and return the response from the output
  return prompter.get_response(output)


In [ ]:
# Old testing code follows.

"""
# testing code for readme
for instruction in [
    "Tell me about alpacas.",
    "Tell me about the president of Mexico in 2019.",
    "Tell me about the king of France in 2019.",
    "List all Canadian provinces in alphabetical order.",
    "Write a Python program that prints the first 10 Fibonacci numbers.",
    "Write a program that prints the numbers from 1 to 100. But for multiples of three print 'Fizz' instead of the number and for the multiples of five print 'Buzz'. For numbers which are multiples of both three and five print 'FizzBuzz'.",  # noqa: E501
    "Tell me five words that rhyme with 'shock'.",
    "Translate the sentence 'I have no mouth but I must scream' into Spanish.",
    "Count up from 1 to 500.",
]:
    print("Instruction:", instruction)
    print("Response:", evaluate(instruction))
    print()
"""


# 1_Classification Task

## dataset

In this section, you'll do a task called NLI (Natural Language Inference) using BERT, LLaMA, and ChatGPT on the MNLI dataset.


---



Start by running the cells below to load the MNLI dataset. The following code will select 200 random samples from the 'mismatched' test dataset. Remember that setting the seed ensures that the sample remains the same in each run.

In [ ]:
!pip install datasets

In [ ]:
from datasets import load_dataset

SEED = 32
N_SAMPLES = 200
mnli_dataset = load_dataset("glue", "mnli")['validation_mismatched']
mnli_samples = mnli_dataset.shuffle(SEED).select(range(N_SAMPLES))

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [ ]:
import torch
# Ensure GPU is available
device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using device:", device)

Using device: cuda


## **BERT**

### 1: Install Required Librarie

In [ ]:
!pip install transformers torch

### 2: Load the Pre-Trained BERT Model

Now, load a BERT model. You can either fine-tune a BERT model for this task or find a pre-finetuned checkpoint on Hugging Face and load it. Please use the `bert-base-uncased`.

In [ ]:
# bert for NLI task

In [ ]:
from transformers import AutoModelForSequenceClassification, AutoTokenizer
import torch

model_name = "JeremiahZ/bert-base-uncased-mnli"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/348 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/712k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/933 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

### 3: Prepare the Data

tokenize the samples from the MNLI dataset to make them compatible with the BERT model

In [ ]:
# tokenizes the 'premise' and 'hypothesis' texts of each example in the dataset.
def tokenize_function(examples):
    return tokenizer(examples["premise"], examples["hypothesis"], padding="max_length", truncation=True, max_length=128)

tokenized_samples = mnli_samples.map(tokenize_function, batched=True)

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

### 4: Making Predictions and Calculating Accuracy

Compute the accuracy of your BERT model for the 200-sample MNLI dataset

In [ ]:
#accuracy

In [ ]:
model.to(device)

def predict(batch):
    # Convert each item in the batch to a tensor and move it to the device
    inputs = {key: torch.tensor(val).to(device) for key, val in batch.items() if key in tokenizer.model_input_names}
    # Model inference without gradient calculation
    with torch.no_grad():
        outputs = model(**inputs)
        # Get the predicted label (index of the max logit)
        predictions = torch.argmax(outputs.logits, dim=-1)
    return {"predictions": predictions.cpu().numpy()}

# Apply the prediction function to the dataset
results = tokenized_samples.map(predict, batched=True)

# Calculate the number of correct predictions and the accuracy
correct_predictions = sum([int(p == l) for p, l in zip(results["predictions"], tokenized_samples["label"])])
accuracy = correct_predictions / len(tokenized_samples)
print(f"Accuracy: {accuracy:.4f}")


Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Accuracy: 0.8150




---



##Llama

>LLaMA (Large Language Model Meta AI) is a family of large language models (LLMs), released by Meta AI starting in February 2023. For the first version of LLaMA, four model sizes were trained: 7, 13, 33 and 65 billion parameters. LLaMA's developers reported that the 13B parameter model's performance on most NLP benchmarks exceeded that of the much larger GPT-3 (with 175B parameters) and that the largest model was competitive with state of the art models such as PaLM and Chinchilla. Whereas the most powerful LLMs have generally been accessible only through limited APIs (if at all), Meta released LLaMA's model weights to the research community under a noncommercial license. In July 2023, Meta released several models as Llama 2, using 7, 13 and 70 billion parameters.[Wiki](https://en.wikipedia.org/wiki/LLaMA)

### Load model

In this section, you'll utilize the LLaMA model to perform the NLI task. Given the server account provided for you, I've already downloaded the model checkpoint, so you just need to load it again. Run the codes below to load the LLaMA model.

Keep in mind that if you choose not to use the server account for this part, you should submit an access request to the model using this [form](https://ai.meta.com/resources/models-and-libraries/llama-downloads/) for Meta models, this [video](https://youtu.be/Z6sCl6abJj4?si=DN428WDxnQ8pUBiF) might be helpful in this matter.



In [ ]:
!pip install --upgrade huggingface_hub

In [ ]:
from huggingface_hub import login
login('XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX') # toker for the course account on huggingface

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [ ]:
from transformers import AutoTokenizer, LlamaForCausalLM
import transformers
import torch

model_name = "meta-llama/Llama-2-7b-chat-hf"
llama_tokenizer = AutoTokenizer.from_pretrained(model_name, use_auth_token=True)
llama_model = LlamaForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16).to(device)

/usr/local/lib/python3.10/dist-packages/transformers/models/auto/tokenization_auto.py:671: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers. Please use `token` instead.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.62k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

### generate answers and claculate accuracy

Now that you have your tokenizer and the model ready you need to desing a suitable promt to perform classification task on the provided 200 samples.

To perform a classification task:

1. Create a clear and informative prompt to guide the generative model in understanding the specific requirements of the classification task.

2. Generate output using the prompt. For LLaMa, utilize the `generate` function.

3. Post-process the generated output to identify the classification format answer. In MNLI, determine if the input falls into categories such as contradiction, entailment, or neutral.

Finally, calculate the accuracy.


In [ ]:
def classify_with_llama(premise, hypothesis):
    prompt = (f"Classify the relationship between the following premise and hypothesis into categories: "
              f"Entailment, Contradiction, or Neutral.\n\n"
              f"Premise: \"{premise}\"\n"
              f"Hypothesis: \"{hypothesis}\"\n\n"
              f"Classification:")

    inputs = llama_tokenizer(prompt, return_tensors='pt').to(device)
    outputs = llama_model.generate(**inputs)
    prediction = llama_tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Clean up the prediction to extract the classification
    return prediction


### accuracy

In [ ]:
correct = 0
total = 0

for sample in mnli_samples:
    premise, hypothesis, label = sample['premise'], sample['hypothesis'], sample['label']
    prediction = classify_with_llama(premise, hypothesis)
    # Extract only the response part (after "Classification:")
    prediction = prediction.split("Classification:")[1].strip().lower().split("\n")[0].strip()

    # Match the prediction with numeric labels
    if 'entailment' in prediction:
        predicted_label = 0
    elif 'contradiction' in prediction:
        predicted_label = 1
    elif 'neutral' in prediction:
        predicted_label = 2
    else:
        predicted_label = -1  # Unknown or unclear prediction

    # Check if the prediction matches the actual label
    if predicted_label == label:
        correct += 1
    total += 1

# Calculate the accuracy
accuracy = correct / total
print(f"Accuracy: {accuracy:.4f}")


Accuracy: 0.3500


### کثیف کاری

In [ ]:
# # model.to(device)

# def classify_with_llama(premise, hypothesis):
#     # Constructing a prompt to evaluate the relationship between premise and hypothesis
#     prompt = f"Premise: {premise}\nHypothesis: {hypothesis}\nIs the hypothesis entailment, contradiction, or neutral relative to the premise?"
#     inputs = tokenizer(prompt, return_tensors='pt')
#     outputs = model.generate(**inputs)
#     prediction = tokenizer.decode(outputs[0], skip_special_tokens=True)
#     return prediction

# # Running the classify_with_llama function for each sample in the MNLI dataset and calculating accuracy
# correct = 0
# for sample in mnli_samples:
#     premise, hypothesis, label = sample['premise'], sample['hypothesis'], sample['label']
#     prediction = classify_with_llama(premise, hypothesis)
#     # Convert prediction to equivalent category
#     if 'entailment' in prediction:
#         predicted_label = 0
#     elif 'contradiction' in prediction:
#         predicted_label = 1
#     elif 'neutral' in prediction:
#         predicted_label = 2
#     else:
#         predicted_label = -1  # Unknown or unclear prediction
#     # Comparing prediction with actual label
#     if predicted_label == label:
#         correct += 1

# accuracy = correct / len(mnli_samples)
# print(f"Accuracy: {accuracy:.4f}")


In [ ]:
# model.to(device)

# def classify_with_llama(premise, hypothesis):
#     prompt = f"Premise: {premise}\nHypothesis: {hypothesis}\nIs the hypothesis entailment, contradiction, or neutral relative to the premise?"
#     inputs = tokenizer(prompt, return_tensors='pt').to(device)
#     outputs = model.generate(**inputs)
#     prediction = tokenizer.decode(outputs[0], skip_special_tokens=True)
#     return prediction
# # Initialize the counter for correct predictions
# correct = 0
# i=0
# n=5
# # Iterate over the MNLI samples
# for sample in mnli_samples[:10]:
#     premise, hypothesis, label = sample['premise'], sample['hypothesis'], sample['label']
#     prediction = classify_with_llama(premise, hypothesis)

#     # Map the prediction to an integer label
#     if 'entailment' in prediction:
#         predicted_label = 0
#     elif 'contradiction' in prediction:
#         predicted_label = 1
#     elif 'neutral' in prediction:
#         predicted_label = 2
#     else:
#         predicted_label = -1  # Unknown or unclear prediction

#     # Check if the prediction matches the actual label
#     if predicted_label == label:
#         correct += 1
#     i+=1
#     if i==n:
#       break

# # Calculate the accuracy
# accuracy = correct / len(mnli_samples)
# print(f"Accuracy: {accuracy:.4f}")


In [ ]:
# def classify_with_llama(premise, hypothesis):
    # Adding a more direct instruction to the prompt for clarity
    # prompt = (f"Premise: {premise}\n"
    #           f"Hypothesis: {hypothesis}\n"
    #           f"Task: Determine if the hypothesis is an entailment, a contradiction, "
    #           f"or neutral in relation to the premise. Answer with 'Entailment', 'Contradiction', or 'Neutral'.")
    # prompt = (f"Premise: {premise}\n"
    #       f"Hypothesis: {hypothesis}\n"
    #       f"Question: Does the hypothesis entail, contradict, or neutralize the premise? "
    #       f"Answer with ONLY 'Entailment', 'Contradiction', or 'Neutral'.")
    # prompt = (f"Read the premise and the hypothesis. Decide whether the hypothesis is an 'entailment', "
    #           f"a 'contradiction', or 'neutral' in relation to the premise. Provide only one word as the answer.\n"
    #           f"Premise: {premise}\nHypothesis: {hypothesis}\nAnswer:")
    # prompt = (f"Read the premise and the hypothesis. Decide whether the hypothesis is 'entailment', "
    #       f"'contradiction', or 'neutral' in relation to the premise. Answer ONLY with one word: 'entailment', 'contradiction', or 'neutral'.\n"
    #       f"Premise: {premise}\nHypothesis: {hypothesis}\nAnswer:")
    # # prompt = f"Premise: {premise}\nHypothesis: {hypothesis}\nIs the hypothesis entailment, contradiction, or neutral relative to the premise?"
    # inputs = tokenizer(prompt, return_tensors='pt').to(device)
    # outputs = model.generate(**inputs)
    # prediction = tokenizer.decode(outputs[0], skip_special_tokens=True)
    # return prediction[len(prompt):]


In [ ]:
# def classify_with_llama(premise, hypothesis):
#     prompt = (f"Classify the relationship between the following premise and hypothesis into categories: "
#               f"Entailment, Contradiction, or Neutral.\n\n"
#               f"Premise: \"{premise}\"\n"
#               f"Hypothesis: \"{hypothesis}\"\n\n"
#               f"Classification:")

#     inputs = tokenizer(prompt, return_tensors='pt').to(device)
#     outputs = model.generate(**inputs)
#     prediction = tokenizer.decode(outputs[0], skip_special_tokens=True)

#     # Clean up the prediction to extract the classification
#     prediction = prediction.split("Classification:")[1].strip().split("\n")[0].strip()
#     return prediction


In [ ]:
# # Initialize the counter for correct predictions
# correct = 0
# i=0
# n=30
# # Iterate over the MNLI samples
# for sample in mnli_samples:
#     premise, hypothesis, label = sample['premise'], sample['hypothesis'], sample['label']
#     prediction = classify_with_llama(premise, hypothesis)
#     # print(prediction)
#     # print('--------------------------------------------------------')

#     # Map the prediction to an integer label
#     if 'entailment' in prediction:
#         predicted_label = 0
#     elif 'contradiction' in prediction:
#         predicted_label = 1
#     elif 'neutral' in prediction:
#         predicted_label = 2
#     else:
#         predicted_label = -1  # Unknown or unclear prediction

#     # Check if the prediction matches the actual label
#     if predicted_label == label:
#         correct += 1
#     i+=1
#     if i==n:
#       break

# # Calculate the accuracy
# # accuracy = correct / len(mnli_samples)
# accuracy = correct / n

# print(f"Accuracy: {accuracy:.4f}")

In [ ]:
# correct = 0
# total = 0
# i=0
# n=10

# for sample in mnli_samples:
#     premise, hypothesis, label = sample['premise'], sample['hypothesis'], sample['label']
#     prediction = classify_with_llama(premise, hypothesis)
#     print(prediction)
#     # Extract only the response part (after "Answer:")
#     # prediction = prediction.split('Answer:')[1].strip().lower()


#     # Match the prediction with numeric labels
#     if 'Entailment' in prediction:
#         predicted_label = 0
#     elif 'Contradiction' in prediction:
#         predicted_label = 1
#     elif 'Neutral' in prediction:
#         predicted_label = 2
#     else:
#         predicted_label = -1  # Unknown or unclear prediction

#     # Check if the prediction matches the actual label
#     if predicted_label == label:
#         correct += 1
#     total += 1
#     print(predicted_label, label)
#     print('---------------------------------')

#     i+=1
#     if i==n:
#       break

# # Calculate the accuracy
# accuracy = correct / total
# print(f"Accuracy: {accuracy:.4f}")


##Chat GPT

You're familiar with ChatGPT and use it daily. However, for more extensive tasks, such as performing the NLI task on our 200-sample dataset, you'll need to utilize its API to automatically obtain answers for the input.

To begin, you'll require an OpenAI account with an authorized phone number to access the API. Once you obtain the token from your account, you can use the provided code to connect to the OpenAI API and access its services. For your convenience, we've created an account with a token that you can use.

In [ ]:
!pip install openai
# !pip install openai==0.28
!pip install --upgrade typing_extensions


In [ ]:
# import openai
# openai.api_key = 'sk-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'

To perform a classification task using ChatGPT, follow the general steps outlined in the previous section. Refer to this [link](https://platform.openai.com/docs/guides/text-generation/chat-completions-api) for instructions on preparing and requesting the model. As mention in the link please use the `gpt-3.5-turbo` model. The pricing information for API usage is also available on this [page](https://openai.com/pricing), in case you're curious.

Keep in mind that there is a limit on API usage, so optimize your requests accordingly.


### requesting GPT model

In [ ]:
from openai import OpenAI

client = OpenAI(
    # Set OpenAI API key
    api_key= "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"
)

def classify_with_chatgpt(premise, hypothesis, model="gpt-3.5-turbo"):
    # Constructing the prompt for ChatGPT
    prompt = (f"'Entailment', 'Contradiction', or 'Neutral'.\n\n"
              f"Premise: \"{premise}\"\n"
              f"Hypothesis: \"{hypothesis}\"\n\n"
              f"What is the classification? ")

    messages = [
        {"role": "system", "content":"Classify the relationship between the following premise and hypothesis into categories"},
        {"role": "user", "content":prompt}
    ]

    response = client.chat.completions.create(
      model=model,
      messages=messages
    )

    # Extracting the text from the response
    prediction = response.choices[0].message.content

    return prediction

# Example usage
premise_example = "A soccer game with multiple males playing."
hypothesis_example = "Some men are playing a sport."
classification = classify_with_chatgpt(premise_example, hypothesis_example)
print("Classification:", classification)


Classification: Entailment


### accuracy

Do not forget to calculate the accuracy!

In [ ]:
def calculate_accuracy(mnli_samples, classify_func):
    correct = 0
    total = 0

    for sample in mnli_samples:
        premise, hypothesis, true_label = sample['premise'], sample['hypothesis'], sample['label']
        predicted_classification = classify_func(premise, hypothesis)

        # Extracting only the classification part of the prediction
        predicted_classification = predicted_classification.lower().strip()

        # Mapping the predicted classification to the corresponding numeric label
        if 'entailment' in predicted_classification:
            predicted_label = 0
        elif 'contradiction' in predicted_classification:
            predicted_label = 1
        elif 'neutral' in predicted_classification:
            predicted_label = 2
        else:
            predicted_label = -1  # Unknown or unclear prediction

        # Comparing the predicted label with the true label
        if predicted_label == true_label:
            correct += 1
        total += 1

    # Calculating the accuracy
    accuracy = correct / total
    return accuracy

# Example usage
accuracy = calculate_accuracy(mnli_samples, classify_with_chatgpt)
print(f"Accuracy: {accuracy:.4f}")


Accuracy: 0.4350


#### کثیف کاری

In [ ]:
# import openai
# model="gpt-3.5-turbo"
# messages = [ {"role": "system",
#               "content":"Classify the relationship between the following premise and hypothesis into categories"} ]

# correct = 0
# total = 0

# for sample in mnli_samples:
# 		premise, hypothesis, true_label = sample['premise'], sample['hypothesis'], sample['label']
# 		# Constructing the prompt for ChatGPT
# 		prompt = (f"'Entailment', 'Contradiction', or 'Neutral'.\n\n"
# 		f"Premise: \"{premise}\"\n"
# 		f"Hypothesis: \"{hypothesis}\"\n\n"
# 		f"What is the classification? ")


# 		messages.append(
# 				{"role": "user", "content":prompt}
# 		)
# 		response = client.chat.completions.create(
# 					model=model,
# 					messages=messages
# 				)

# 		reply = response.choices[0].message.content
# 		messages.append({"role": "assistant", "content": reply})
# 		predicted_classification = reply.lower().strip()

# 		# Mapping the predicted classification to the corresponding numeric label
# 		if 'entailment' in predicted_classification:
# 				predicted_label = 0
# 		elif 'contradiction' in predicted_classification:
# 				predicted_label = 1
# 		elif 'neutral' in predicted_classification:
# 				predicted_label = 2
# 		else:
# 				predicted_label = -1  # Unknown or unclear prediction

# 		# Comparing the predicted label with the true label
# 		if predicted_label == true_label:
# 				correct += 1
# 		total += 1

# # Calculating the accuracy
# accuracy = correct / total

# print(f"Accuracy: {accuracy:.4f}")


## Interpretation of results

In conclusion, compare the results from all three models and discuss the obtained accuracies.

Based on the accuracy scores the BERT, LLaMA, and ChatGPT (GPT-3.5 turbo) models on the Natural Language Inference (NLI) task using the MNLI dataset, we can derive several insights and interpretations:

1. **BERT's Performance (0.8150 Accuracy)**:
   - BERT's high accuracy indicates its strong performance on structured NLP tasks like NLI.
   - This result is likely due to BERT's architecture and training, which is specifically designed to understand the context and relationships between sentences.
   - The fine-tuning process of BERT on a specific task like NLI can significantly improve its performance, as evidenced by the accuracy score.

2. **Llama's Performance (0.3500 Accuracy)**:
   - The comparatively lower accuracy of LLaMA suggests challenges in handling the NLI task.
   - This could be due to several factors such as the version of the LLaMA model used (if it's a smaller version, it might not perform as well), or it might not have been specifically fine-tuned for the NLI task.
   - The difference in model architecture and training objectives between BERT and LLaMA might also play a role in this performance gap.

3. **ChatGPT (GPT-3.5 turbo) Performance (0.4350 Accuracy)**:
   - ChatGPT's accuracy, while better than LLaMA's, is still significantly lower than BERT's.
   - This could be because GPT-3.5, is a more generalized model and not specifically optimized for NLI tasks.
   - The broader training and objective of GPT models, which aim to generate human-like text, might not align perfectly with the precision required for NLI tasks.

4. **Comparative Analysis**:
   - The differences in accuracies highlight the importance of model architecture and training objectives.
   - BERT, being a bidirectional transformer specifically fine-tuned for NLI, excels in this task.
   - LLaMA and GPT-3.5, despite being powerful language models, may not be as effective for specialized tasks like NLI without specific fine-tuning or adaptations.

5. **Implications for NLP Applications**:
   - These results underscore the need to choose the right model for specific NLP tasks.
   - For tasks requiring understanding of context and relationships between sentences, models like BERT are preferable.
   - In contrast, for more general language generation tasks, models like LLaMA and GPT-3.5 might be more suitable.


---

# 2_Temperature

## theoretical explanation

>The **temperature** parameter is an important hyperparameter in generative language models that can be used to control the randomness and creativity of the generated text.

Search about this hyperparameter and explain how it works? Where is it used in generative models? Also, include the formula for the part of the model that has the temperature parameter and describe how it affects the creativity of the model based on it. Please provide a theoretical explanation for your response.

The **temperature** parameter is a crucial hyperparameter in generative language models, playing a significant role in modulating the randomness and creativity of the text they generate. This parameter essentially adjusts the probability distribution used by the model to select the next word or token in a sequence.

In technical terms, the temperature modifies the logits (i.e., the raw output of the model's last neural layer before applying the softmax function) of the model's predictions. The softmax function then converts these logits into a probability distribution over all possible tokens.

The formula incorporating the temperature parameter $\tau$ is as follows:

$$ P(i) = \frac{e^{(logit(i)/\tau)}}{\sum_j e^{(logit(j)/\tau)}} $$

Where $P(i)$ is the probability of token $i$, and $logit(i)$ is the logit for token $ i $. The sum in the denominator runs over all possible tokens $ j $.

The temperature parameter $\tau $ influences this distribution in the following ways:

1. **Low Temperature ($\tau < 1\$)**: This setting sharpens the probability distribution, making the model more confident and less random in its predictions. It increases the likelihood of the model choosing more probable tokens, leading to more predictable and conservative text generation.

2. **High Temperature ($\tau > 1$)**: Conversely, a high temperature smooths the probability distribution, making the model less confident in its predictions. This results in a higher likelihood of selecting less probable tokens, thereby introducing more randomness and creativity into the generated text.

In generative models, the temperature is used to balance between coherency and creativity. A lower temperature is typically preferred for tasks requiring high precision and factual correctness, while a higher temperature may be desired in scenarios where creativity and diversity of expression are more valued.

By carefully adjusting the temperature, users can control the style and characteristics of the generated text, making this parameter a powerful tool in the hands of a skilled practitioner in natural language generation.


---

## Practical experimentally

For both generative models in this assignment, you have the option to adjust the temperature hyper-parameter when generating the output. For the given prompts, explore the responses of *Llama* and *ChatGPT* models. use five different temperature settings that cover both low and high ranges.


In [ ]:
prompt1 = 'Once upon a time'
prompt2 = 'In a world where cats can speak, the first thing a cat said was:'

# Define temperature settings
temperatures = [0.2, 0.4, 0.6, 0.8, 1.0]

### Llama

In [ ]:
def generate_llama_responses(prompt, temperatures, model=llama_model, tokenizer=llama_tokenizer):
    """
    Generate responses for a given prompt using different temperature settings with the Llama model.

    param prompt: The prompt to generate responses for.
    param temperatures: A list of temperature settings to use.
    param model: The Llama model.
    param tokenizer: The tokenizer for the Llama model.
    return: A dictionary with temperature as keys and generated responses as values.
    """
    responses = {}
    for temp in temperatures:
        # Encode the prompt
        input_ids = tokenizer.encode(prompt, return_tensors="pt").to(device)

        # Generate the response
        outputs = model.generate(input_ids, max_length=50, temperature=temp, num_return_sequences=1)
        response = tokenizer.decode(outputs[0], skip_special_tokens=True)

        responses[temp] = response
    return responses

# Generate responses for both prompts
responses_prompt1 = generate_llama_responses(prompt1, temperatures)
responses_prompt2 = generate_llama_responses(prompt2, temperatures)

# Print the responses
print("Responses for Prompt 1 with Llama:")
for temp, response in responses_prompt1.items():
    print(f"Temperature {temp}: {response}")

print("\nResponses for Prompt 2 with Llama:")
for temp, response in responses_prompt2.items():
    print(f"Temperature {temp}: {response}")


Responses for Prompt 1 with Llama:
Temperature 0.2: Once upon a time, in a far-off kingdom, there lived a young prince named Leo. Leo was the eldest son of the king and queen, and he was known throughout the kingdom for his kindness, bravery, and wisdom
Temperature 0.4: Once upon a time, in a far-off land, there was a magical kingdom ruled by a wise and just king. The kingdom was known for its beauty and prosperity, and the king was loved by all his subjects.


Temperature 0.6: Once upon a time, in a far-off land, there was a magical forest filled with towering trees, sparkling streams, and a vast array of creatures. Some of these creatures were friendly and kind, while others were
Temperature 0.8: Once upon a time, in a small village nestled in the rolling hills of the countryside, there lived a little girl named Emily.
Emily was a curious and adventurous child, always eager to explore the world
Temperature 1.0: Once upon a time, there was a farmer who lived in a small village. Here h

### Chat GPT

In [ ]:
def generate_chatgpt_responses(prompt, temperatures):
    """
    Generate responses for a given prompt using different temperature settings.

    :param prompt: The prompt to generate responses for.
    :param temperatures: A list of temperature settings to use.
    :return: A dictionary with temperature as keys and generated responses as values.
    """
    responses = {}
    for temp in temperatures:
        response = client.completions.create(
            model="gpt-3.5-turbo-instruct",
            prompt=prompt,
            temperature=temp,
            max_tokens=50
        )
        responses[temp] = response.choices[0].text.strip()
    return responses

# Generate responses for both prompts
responses_prompt1 = generate_chatgpt_responses(prompt1, temperatures)
responses_prompt2 = generate_chatgpt_responses(prompt2, temperatures)

# Print the responses
print("Responses for Prompt 1 with ChatGPT:")
for temp, response in responses_prompt1.items():
    print(f"Temperature {temp}: {response}")

print("\nResponses for Prompt 2 with ChatGPT:")
for temp, response in responses_prompt2.items():
    print(f"Temperature {temp}: {response}")


Responses for Prompt 1 with ChatGPT:
Temperature 0.2: , in a faraway land, there lived a young prince named Alexander. He was the only son of the king and queen, and was loved by all in the kingdom for his kind heart and brave spirit.

One day, while exploring the forest near
Temperature 0.4: , there was a little girl named Goldilocks. She lived in a small village at the edge of a deep, dark forest. Goldilocks was a curious and adventurous girl, and she loved to explore the forest and all its wonders.

One
Temperature 0.6: , I was one of those people who thought that "getting fit" meant "getting thin." And, in fact, that's why I started running. I had packed on a few pounds between college and my wedding, and I wanted to get back
Temperature 0.8: , I was a girl who was in love with a boy. We were both in high school and had been friends since we were kids. I always had a crush on him but never had the courage to tell him. It wasn't until our senior
Temperature 1.0: , a hand grasping a 

### Interpretation of results

Discuss the results and compare the outputs for different temperatures. Do the outcomes align with your expectations based on the definition of temperature?

Enjoy!!

The results from the LLaMA and ChatGPT models at different temperature settings offer an interesting perspective on how temperature influences language model outputs.

#### LLaMA Model Responses:

1. **Lower Temperatures (0.2 and 0.4)**:
   - The responses are more structured and conventional.
   - The narratives start with traditional fairy tale openings, suggesting a tendency towards more predictable and safer outputs.

2. **Medium Temperature (0.6)**:
   - There's a noticeable shift towards more creative and less conventional storytelling.
   - The introduction of a magical forest and a range of creatures shows increased creativity.

3. **Higher Temperatures (0.8 and 1.0)**:
   - The responses become increasingly divergent and less predictable.
   - At the highest temperature (1.0), the narrative shifts significantly, moving away from traditional fairy tales to a more mundane and relatable scenario.

#### ChatGPT (GPT-3.5 turbo) Model Responses:

1. **Lower Temperatures (0.2 and 0.4)**:
   - The narratives are coherent and follow a traditional storyline.
   - The focus is on familiar fairy tale and adventure themes, indicating a preference for safe and conventional outputs.

2. **Medium Temperature (0.6)**:
   - The content starts to deviate from traditional storytelling, reflecting a more personal and reflective narrative.
   - This indicates a balance between creativity and coherence.

3. **Higher Temperatures (0.8 and 1.0)**:
   - The stories become less conventional and more diverse in themes.
   - The highest temperature setting leads to unexpected and unconventional outputs, such as the focus on a pizza slice, showing increased randomness.

#### General Observations:

- **Impact of Temperature**: As the temperature increases, both models demonstrate a shift from traditional, predictable narratives to more creative, diverse, and sometimes random outputs. This aligns with the theoretical understanding of temperature in language models, where lower temperatures lead to more predictable and conservative outputs, and higher temperatures result in more creativity and randomness.
- **Model Differences**: There's a noticeable difference in how LLaMA and ChatGPT handle different temperatures. LLaMA seems to retain a fairy tale-like narrative structure across temperatures, whereas ChatGPT shows a broader range of themes and narratives.
- **Balancing Creativity and Coherence**: Medium temperatures (around 0.6) seem to strike a balance between maintaining coherence and introducing creative elements in the narratives for both models.
- **Randomness at High Temperatures**: At the highest temperature settings, both models exhibit a significant increase in randomness and deviation from conventional narratives. This indicates that while high temperatures can introduce creative and novel elements, they can also lead to less coherent and predictable outputs.

These insights demonstrate the nuanced effects of temperature on language model outputs and highlight the importance of carefully selecting temperature settings based on the desired balance between creativity and predictability in generated text.